### PR TITLE
T6246: basic haproxy http-check configuration

### PR DESCRIPTION
**Pull request by nvollmar**
Enable http-check configuration for load-balancing revers-proxy.
I'm pretty new to Vyos and this is my first attempt to contribute and hope to get some feedback

## Change Summary
Allow configuration of http-check for haproxy backends

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related PR(s)
vyos/vyos-documentation#1388

## Component(s) name
load-balancing

## Proposed changes
Enable configuration options for haproxy backend http-check

## How to test
```
set load-balancing reverse-proxy backend master http-check method 'get'
set load-balancing reverse-proxy backend master http-check uri '/health'
set load-balancing reverse-proxy backend master http-check expect 'status 200'
```

## Smoketest result
```
test_01_lb_reverse_proxy_domain (__main__.TestLoadBalancingReverseProxy.test_01_lb_reverse_proxy_domain) ... ok
test_02_lb_reverse_proxy_cert_not_exists (__main__.TestLoadBalancingReverseProxy.test_02_lb_reverse_proxy_cert_not_exists) ... 
PKI does not contain any certificates!


Certificate "cert" not found in configuration!

ok
test_03_lb_reverse_proxy_ca_not_exists (__main__.TestLoadBalancingReverseProxy.test_03_lb_reverse_proxy_ca_not_exists) ... ok
test_04_lb_reverse_proxy_backend_ssl_no_verify (__main__.TestLoadBalancingReverseProxy.test_04_lb_reverse_proxy_backend_ssl_no_verify) ... 
backend bk-01 cannot have both ssl options no-verify and ca-certificate
set!

ok
test_05_lb_reverse_proxy_backend_http_check (__main__.TestLoadBalancingReverseProxy.test_05_lb_reverse_proxy_backend_http_check) ... ok

----------------------------------------------------------------------
Ran 5 tests in 76.137s

OK
```

## Checklist:
- [X] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [X] I have linked this PR to one or more Phabricator Task(s)
- [X] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [X] My commit headlines contain a valid Task id
- [X] My change requires a change to the documentation
- [x] I have updated the documentation accordingly

### CODE REVIEW

File: interface-definitions/load-balancing_reverse-proxy.xml.in [hunk 1]
```diff
@@ -102,6 +102,45 @@
...
+                  <help>HTTP check configuration</help>
+                  <completionHelp>
+                    <list>uri /health</list>
+                  </completionHelp>
+                </properties>
+                <children>
+                  <leafNode name="method">
+                    <properties>
+                      <valueHelp>
+                        <format>OPTIONS|GET|POST|PUT</format>
```
Comment (dmbaturin) [inline interface-definitions/load-balancing_reverse-proxy.xml.in]: We do not use uppercase identifiers in the CLI. CLI options should be lowercase in the config and the config script or the template can uppercase them for the HAProxy config.

**Review by dmbaturin — CHANGES_REQUESTED** [review]
Please make sure to read and follow CONTRIBUTING.md (rather than just check the checkbox ;).

You need to create a task in https://vyos.dev and reference its id in the PR title and the commit message. Without linking commits to tasks it's very hard to prepare any release notes afterwards.

The idea looks good to me, just needs to follow the guidelines.

File: interface-definitions/load-balancing_reverse-proxy.xml.in (hunk 1 above)
Comment (nvollmar) [reply interface-definitions/load-balancing_reverse-proxy.xml.in]: Thanks for the hints, I've updated the method node accordingly.

I've also applied for an account in vyos.dev to create a task. 
Just found a post in the forum how to run the smoketests, so I'll try that next.

File: smoketest/scripts/cli/test_load-balancing_reverse-proxy.py [hunk 2]
```diff
@@ -298,6 +298,21 @@ def test_04_lb_reverse_proxy_backend_ssl_no_verify(self):
         with self.assertRaises(ConfigSessionError) as e:
             self.cli_commit()
 
+    def test_05_lb_reverse_proxy_backend_http_check(self):
+      # Setup base
+      self.base_config()
+
+      # Set http-check
+      self.cli_set(base_path + ['backend', 'bk-01', 'http-check', 'method', 'GET'])
```
Comment (sever-sever) [inline smoketest/scripts/cli/test_load-balancing_reverse-proxy.py]: This test is going to fail, as the accepted format `get`
```
<format>options|get|post|put</format>
```

File: interface-definitions/load-balancing_reverse-proxy.xml.in [hunk 3]
```diff
@@ -102,6 +102,45 @@
                   </leafNode>
                 </children>
               </node>
+              <tagNode name="http-check">
```
Comment (sever-sever) [inline interface-definitions/load-balancing_reverse-proxy.xml.in]: Why is this the tagNode?
```
set load-balancing reverse-proxy backend back http-check method GET
set load-balancing reverse-proxy backend back http-check uri /health

```
As I see it should be `node`

**Review by sever-sever — CHANGES_REQUESTED** [review]

File: interface-definitions/load-balancing_reverse-proxy.xml.in (hunk 3 above)
Comment (nvollmar) [reply interface-definitions/load-balancing_reverse-proxy.xml.in]: Thanks, fixed that.

File: interface-definitions/load-balancing_reverse-proxy.xml.in [hunk 4]
```diff
@@ -102,6 +102,45 @@
                   </leafNode>
                 </children>
               </node>
+              <node name="http-check">
+                <properties>
+                  <help>HTTP check configuration</help>
+                  <completionHelp>
+                    <list>uri /health</list>
+                  </completionHelp>
```
Comment (sever-sever) [inline interface-definitions/load-balancing_reverse-proxy.xml.in]: It is not required and looks strange, as you use the next leafnodes `method` and `uri`

File: interface-definitions/load-balancing_reverse-proxy.xml.in [hunk 5]
```diff
@@ -102,6 +102,45 @@
...
+                </properties>
+                <children>
+                  <leafNode name="method">
+                    <properties>
+                      <valueHelp>
```
Comment (sever-sever) [inline interface-definitions/load-balancing_reverse-proxy.xml.in]: Should there be an additional `help`?

**Review by sever-sever — CHANGES_REQUESTED** [review]

**Comment by sever-sever** [comment]
Show please the set of the full config and generated haproxy config.

File: interface-definitions/load-balancing_reverse-proxy.xml.in (hunk 4 above)
Comment (nvollmar) [reply interface-definitions/load-balancing_reverse-proxy.xml.in]: removed

**Comment by nvollmar** [comment]
I've built an iso with those changes and this is the config entered:
```
set load-balancing reverse-proxy service api listen-address '10.3.0.1'
set load-balancing reverse-proxy service api port '6443'
set load-balancing reverse-proxy service api mode 'tcp'
set load-balancing reverse-proxy service api backend 'master'
set load-balancing reverse-proxy backend master http-check method 'get'
set load-balancing reverse-proxy backend master http-check uri '/health'
set load-balancing reverse-proxy backend master http-check expect 'status 200'
set load-balancing reverse-proxy backend master mode 'tcp'
set load-balancing reverse-proxy backend master balance 'round-robin'
set load-balancing reverse-proxy backend master server master-1 address '10.3.1.1'
set load-balancing reverse-proxy backend master server master-1 port '6443'
set load-balancing reverse-proxy backend master server master-1 check
set load-balancing reverse-proxy backend master server master-2 address '10.3.1.2'
set load-balancing reverse-proxy backend master server master-2 port '6443'
set load-balancing reverse-proxy backend master server master-2 check
```

and this is the resulting configuration:
```

vyos@vyos:/$ cat ./run/haproxy/haproxy.cfg
### Autogenerated by load-balancing_reverse-proxy.py ###

global
    log /dev/log local0
    log /dev/log local1 notice
    chroot /var/lib/haproxy
    stats socket /run/haproxy/admin.sock mode 660 level admin
    stats timeout 30s
    user haproxy
    group haproxy
    daemon


    # Default SSL material locations
    ca-base /etc/ssl/certs
    crt-base /etc/ssl/private

    # https://ssl-config.mozilla.org/#server=haproxy&version=2.6.12-1&config=intermediate&openssl=3.0.8-1&guideline=5.6
    ssl-default-bind-ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHH
    ssl-default-bind-ciphersuites TLS_AES_128_GCM_SHA256:TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256
    ssl-default-bind-options force-tlsv13

defaults
    log     global
    mode    http
    option  dontlognull
    timeout connect 10s
    timeout client  50s
    timeout server  50s
    errorfile 400 /etc/haproxy/errors/400.http
    errorfile 403 /etc/haproxy/errors/403.http
    errorfile 408 /etc/haproxy/errors/408.http
    errorfile 500 /etc/haproxy/errors/500.http
    errorfile 502 /etc/haproxy/errors/502.http
    errorfile 503 /etc/haproxy/errors/503.http
    errorfile 504 /etc/haproxy/errors/504.http

# Frontend
frontend api
    bind 10.3.0.1:6443  
    mode tcp
    default_backend master


# Backend
backend master
    option httpchk
    http-check send meth GET uri /health
    http-check expect status 200
    balance roundrobin
    mode tcp
    server master-1 10.3.1.1:6443 check 
    server master-2 10.3.1.2:6443 check 
```

**Comment by nvollmar** [comment]
But I have to remove the default values, otherwise the check will be configured in any case

**Comment by nvollmar** [comment]
I've run the smoketest on the latest change and verified the resulting config again.
I'll prepare a pull request for the configuration next.

File: interface-definitions/load-balancing_reverse-proxy.xml.in [hunk 6]
```diff
@@ -102,6 +102,39 @@
...
+                <children>
+                  <leafNode name="method">
+                    <properties>
+                      <help>HTTP method used for health check</help>
+                      <valueHelp>
+                        <format>options|get|post|put</format>
+                        <description>HTTP method used for health checking</description>
+                      </valueHelp>
+                      <constraint>
+                        <regex>options|get|post|put</regex>
```
Comment (sever-sever) [inline interface-definitions/load-balancing_reverse-proxy.xml.in:118]: What does “options” do? Is it correct method?

Comment (nvollmar) [reply interface-definitions/load-balancing_reverse-proxy.xml.in:118]: HTTP OPTIONS request are used to query available request methods. haproxy uses that per default:

> An HTTP-layer health check sends an HTTP OPTIONS request to the server and expects to get a successful response. To enable it, add option httpchk to the backend section

It is used by default as OPTIONS requests should not trigger any intensive computation like a GET might (to render an HTML page for example).

**Review by sever-sever — APPROVED** [review]

**Comment by dmbaturin** [comment]
@mergifyio backport sagitta

**Pull request merged**